### DISCUSSION TITLE
What format requirements do we need to establish matrix in MATLAB for “ImportMatricesFromMatlab” module?

### DISCUSSION THREAD

**Discussion by zhifeiji-han** [SCIInstitute/SCIRun#2315] [General]
I want to bulid two matrices in the “ImportMatricesFromMatlab” module, and put this matrices into "ReportColumnMatrixMisfit" module,and get the root mean square error(RMSE).
![image](https://user-images.githubusercontent.com/70640243/120650810-bb43c100-c4b0-11eb-9d5b-311b14385612.png)
Here, I have two questions.
First， What format requirements do I need to establish matrices in MATLAB for “ImportMatricesFromMatlab” module?
Secondly, Dose "Rel RMS" in "ReportColumnMatrixMisfit" module mean root mean square error(RMSE)?

**Comment by zhifeiji-han** [comment]
![image](https://user-images.githubusercontent.com/70640243/120764711-05798080-c54b-11eb-8940-ef26d0b59173.png)
The meaning of the parameters in the module is not clear

**Comment by jessdtate** [comment]
`Error Metric` -- the value that will sent to the output port.
`p-value` -- parameter for the p norm option
`Rel RMS` -- relative RMS error
`1-CC` -- inverse correlation coefficient, or 1 minus the correlation.  

This module is really built to accompany a specific opimization module that hasn't been ported to SR5 yet.  

Sounds like you want the RMS error, not the relative RMS.  It might be easiest to use the InterfaceWithPython.  There are instructions on the Python interface [here](http://sciinstitute.github.io/SCIRun/python.html). 

You can get the norm of the difference, a similar value,  using EvaluateLinearAlgebraBinary to get the difference and ReportMatrixSliceMeasure to get the norm of each row/column of that matrix.  

Getting the RMS error is easiest if they are vectors assigned to fields, then the CalculateFieldData module could be used.

**Comment by zhifeiji-han** [comment]
thank you for your detailed explanation!
I want to test the data from MATLAB, need to use  "ImportMatricesFromMatlab" module and "ImportFieldsFromMatlab "module. In particular, "ImportFieldsFromMatlab "module can be used to output multiple matrices. 
Could you explain the data in MATLAB need what format can be normally imported into these two modules ? If you give an example, it will be clearer.

**Comment by zhifeiji-han** [comment]
I try to use "InterfaceWithPython" module ,and RMSE can be calculated.
However, there is a small problem. The RMSE calculated is assigned to matrixout1, and the error is displayed as "input data required on port inputmatrix".
![image](https://user-images.githubusercontent.com/70640243/121156027-ee080380-c87a-11eb-91eb-110465e12fc3.png)
![image](https://user-images.githubusercontent.com/70640243/121156091-fa8c5c00-c87a-11eb-8214-6e9fa64c1b28.png)

**Comment by zhifeiji-han** [comment]
What format requirements do we need to establish matrix in MATLAB for “ImportMatricesFromMatlab” module? 
Please explain the format of the matrix！

**Comment by zhifeiji-han** [comment]
@jessdtate @dcwhite

**Comment by jessdtate** [comment]
I can't read the text in the InterfaceWithPython module.  If you past the code that you have in there, perhaps I can help with that

**Comment by jessdtate** [comment]
the InterfaceWithMatlab module just takes matrices saved in a matlab format and allows you to select them from a list.  The only caveat is that the matrices have to be 2D, and they can't be cell arrays or part of a structure array.  You can save these out from SCIRun, Matlab, or Python with scipy.io. 

 If you are still having trouble, perhaps you could include a small example file that isn't working for you.

**Comment by zhifeiji-han** [comment]
I try to build and save the matrix in MATLAB, and then select in the list, but I don't know how to select in the list. Can you introduce how to make the matrix select in the list?
I have a small example. 
I  create a mat format file in MATLAB, which contains one matrix, A = [1,2,3; 4,5,6]. The  matrices is imported into SCIRun through the "ImportMatricesFromMatlab" module.
The problem I encountered was that the matrix I created was not displayed in the "matlab object" of the "ImportMatricesFromMatlab" module's interface.
![image](https://user-images.githubusercontent.com/70640243/122168287-aef73500-ceae-11eb-8248-2a232e4f9459.png)
![image](https://user-images.githubusercontent.com/70640243/122168359-c33b3200-ceae-11eb-9e55-0e8bbd1482eb.png)

[test_ImportMatricesFromMatlab.zip](https://github.com/SCIInstitute/SCIRun/files/6660694/test_ImportMatricesFromMatlab.zip)

**Comment by zhifeiji-han** [comment]
For the InterfaceWithPython module， I also have an example program，whose code is as follows.
![image](https://user-images.githubusercontent.com/70640243/122173038-ec11f600-ceb3-11eb-819a-0a9b046a0769.png)
![image](https://user-images.githubusercontent.com/70640243/122173080-f9c77b80-ceb3-11eb-9ac3-d632506d6d71.png)
[test_RMSE(python code).zip](https://github.com/SCIInstitute/SCIRun/files/6660882/test_RMSE.python.code.zip)

**Comment by jessdtate** [comment]
Looks like you save the matlab matrix in an HDF format, which isn't supported in SCIRun.  Try saving it with the save function with normal compression.

**Comment by jessdtate** [comment]
Using numpy to compute the RMSE is fine, you'll just need to convert the matrix back to a list to send it back to SCIRun.  Line 15 should be:
```
matrixOutput1 = RMSE.tolist()
```
since this case is a single value, you'll need double brackets to make it a list of lists. 
```
matrixOutput1 = [[RMSE.tolist()]]
```

**Comment by zhifeiji-han** [comment]
Thank you for your detailed reply. Both problems have been solved.
I modified the format of the mat file in the MATLAB software, and the mat file can be imported into the "ImportMatricesFromMatlab"  module normally.
I think the reason may be that the saved version is too high, so an error is reported.
